rack-protection/lib/rack/protection/host_authorization.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'rack/protection'
+require 'ipaddr'
+
+module Rack
+  module Protection
+    ##
+    # Prevented attack::   DNS rebinding and other Host header attacks
+    # Supported browsers:: all
+    # More infos::         https://en.wikipedia.org/wiki/DNS_rebinding
+    #                      https://portswigger.net/web-security/host-header
+    #
+    # Blocks HTTP requests with an unrecognized hostname in any of the following
+    # HTTP headers: Host, X-Forwarded-Host, Forwarded
+    #
+    # If you want to permit a specific hostname, you can pass in as the `:permitted_hosts` option:
+    #
+    #     use Rack::Protection::HostAuthorization, permitted_hosts: ["www.example.org", "sinatrarb.com"]
+    #
+    # The `:allow_if` option can also be set to a proc to use custom allow/deny logic.
+    class HostAuthorization < Base
+      PORT_REGEXP = /:\d+\z/.freeze
+      default_reaction :deny
+      default_options allow_if: nil,
+                      message: "Host not permitted"
+
+      def initialize(*)
+        super
+        @all_permitted_hosts = Array(options[:permitted_hosts])
+        @permitted_hosts = @all_permitted_hosts
+          .select { |host| host.is_a?(String) }
+          .map(&:downcase)
+        @domain_hosts = @permitted_hosts.select { |host| host[0] == "." }
+        @ip_hosts = @all_permitted_hosts.select { |host| host.is_a?(IPAddr) }
+      end
+
+      def accepts?(env)
+        return true if options[:allow_if]&.call(env)
+        return true if @all_permitted_hosts.empty?
```

Comment:
Haven't thought this suggestion through, but just to get the conversation going:

```suggestion
        return true if @permitted_hosts.empty && @ip_hosts.empty?
```

I am not sure about what should happen if one sets `permitted_hosts` to something like `[1, true]`?

Reply:
I think it is more correct and expected that it will never match and requests are blocked (current behaviour).

However, this prompted we to add debug logging: 801e8210 – so it is easier to troubleshoot `HostAuthorization`

Reply:
Yes, makes sense.